File: Scala/src/main/scala/TextbookTrack/Chapter04/BA4H.scala
package TextbookTrack.Chapter04

object BA4H {
  import scala.language.implicitConversions
  import scala.collection.mutable.{Map => MutableMap}

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private implicit def toMassCounts(masses: List[Int]): Map[Int, Int] =
    masses.groupBy(identity).view.mapValues(_.length).toMap

  def calcConvolutionOfSpectrum(spectrum: Map[Int, Int]): List[Int] = {
    val convolution: MutableMap[Int, Int] = MutableMap()
    for {
      (mass1, count1) <- spectrum
      (mass2, count2) <- spectrum
      difference: Int = mass1 - mass2
      if difference > 0
    } convolution(difference) = convolution.getOrElse(difference, 0) + count1 * count2

    convolution
      .toList
      .sortBy{ case (_, count) => count }(Ordering[Int].reverse)
      .flatMap{ case (massDifference, count) => List.fill(count)(massDifference) }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val spectrum: Map[Int, Int] = convertToIntList(reader.next())
    val result: List[Int] = calcConvolutionOfSpectrum(spectrum)
    println(result.mkString(" "))
  }
}

File: Python/textbook_track/chapter04/ba4h.py
# Generate the Convolution of a Spectrum
import itertools as it
import sys

from collections import Counter, defaultdict


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def sort_convolution(convolution):
    sorted_convolution = sorted(convolution.items(), key=lambda x: x[1], reverse=True)
    return list(it.chain.from_iterable(map(lambda mass_counts: it.repeat(*mass_counts), sorted_convolution)))


def calc_convolution_of_spectrum(spectrum):
    convolution = defaultdict(int)
    for mass1, count1 in spectrum.items():
        for mass2, count2 in spectrum.items():
            difference = mass1 - mass2
            if difference > 0:
                convolution[difference] += count1 * count2
    return sort_convolution(dict(convolution))


def main():
    data = sys.stdin.read().splitlines()
    spectrum = Counter(convert_to_intlist(data[0]))
    result = calc_convolution_of_spectrum(spectrum)
    print(' '.join(map(str, result)))


if __name__ == '__main__':
    main()
